### ISSUE TITLE
Not able to compile on windows following the instructions - did I miss something?

### ISSUE THREAD

**Issue by petewil0** [tlemo/cuda_nbody#1]
I have the CUDA_PATH set, but when I do cmake, it doesn't find the CUDA toolset:

c:\Users\petew\source\repos\cuda_nbody\build>echo %CUDA_PATH%
C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v12.5

CMake Error at C:/Program Files/CMake/share/cmake-3.29/Modules/CMakeDetermineCompilerId.cmake:563 (message):
  No CUDA toolset found.
Call Stack (most recent call first):
  C:/Program Files/CMake/share/cmake-3.29/Modules/CMakeDetermineCompilerId.cmake:8 (CMAKE_DETERMINE_COMPILER_ID_BUILD)
  C:/Program Files/CMake/share/cmake-3.29/Modules/CMakeDetermineCompilerId.cmake:53 (__determine_compiler_id_test)
  C:/Program Files/CMake/share/cmake-3.29/Modules/CMakeDetermineCUDACompiler.cmake:131 (CMAKE_DETERMINE_COMPILER_ID)
  CMakeLists.txt:15 (project)

-- Configuring incomplete, errors occurred!

Did I miss something, or do the instructions maybe need a minor update?

**Comment by tlemo** [comment]
1. Which version(s) of Visual Studio do you have installed?
2. Did you install CUDA before, or after installing Visual Studio?

If you installed CUDA before VS, I'd try reinstalling the CUDA SDK.

PS. see also: https://gitlab.kitware.com/cmake/cmake/-/issues/24183

**Comment by petewil0** [comment]
VS2022
I installed VS2022 first, then CUDA.

I used a Developer command prompt.  I also tried a native tools command prompt, same error 

I reinstalled the CUDA sdk and ran again from a developer command prompt and got the same error  (this time I included the full output in case there are other clues)


c:\Users\petew\source\repos\cuda_nbody\build>cmake .. -DCMAKE_BUILD_TYPE=Release
-- Running vcpkg install
Detecting compiler hash for triplet x64-windows...
Compiler found: C:/Program Files (x86)/Microsoft Visual Studio/2022/BuildTools/VC/Tools/MSVC/14.39.33519/bin/Hostx64/x64/cl.exe
All requested packages are currently installed.
Total install time: 200 ns
argparse provides CMake targets:

  # this is heuristically generated, and may not be correct
  find_package(argparse CONFIG REQUIRED)
  target_link_libraries(main PRIVATE argparse::argparse)

The package glew is compatible with built-in CMake targets:

    find_package(GLEW REQUIRED)
    target_link_libraries(main PRIVATE GLEW::GLEW)

glfw3 provides CMake targets:

  # this is heuristically generated, and may not be correct
  find_package(glfw3 CONFIG REQUIRED)
  target_link_libraries(main PRIVATE glfw)

glfw3 provides pkg-config modules:

  # A multi-platform library for OpenGL, window and input
  glfw3

imgui provides CMake targets:

  # this is heuristically generated, and may not be correct
  find_package(imgui CONFIG REQUIRED)
  target_link_libraries(main PRIVATE imgui::imgui)

-- Running vcpkg install - done
-- Selecting Windows SDK version 10.0.22621.0 to target Windows 10.0.22631.
**CMake Error at C:/Program Files/CMake/share/cmake-3.29/Modules/CMakeDetermineCompilerId.cmake:563 (message):
  No CUDA toolset found.**
Call Stack (most recent call first):
  C:/Program Files/CMake/share/cmake-3.29/Modules/CMakeDetermineCompilerId.cmake:8 (CMAKE_DETERMINE_COMPILER_ID_BUILD)
  C:/Program Files/CMake/share/cmake-3.29/Modules/CMakeDetermineCompilerId.cmake:53 (__determine_compiler_id_test)
  C:/Program Files/CMake/share/cmake-3.29/Modules/CMakeDetermineCUDACompiler.cmake:131 (CMAKE_DETERMINE_COMPILER_ID)
  CMakeLists.txt:15 (project)


-- Configuring incomplete, errors occurred!

**Comment by tlemo** [comment]
Sorry for the trouble. Just to double check a few more things:
1. are you using a x64 (amd64) MSVC command prompt, right?
2. you're using the makefile generator, not Visual Studio solution generators?
3. did you install the CUDA Visual Studio Extension part of the CUDA SDK? (this should not be required for makefiles, but it is for the VS solution generators)

If it still doesn't work, please attach the `CMakeFiles/CMakeConfigureLog.yaml` (from under the build directory)

**Comment by petewil0** [comment]
It's OK, thanks for responding!  NO need to apologize, this is how software development works :-)

1. I think I am using a x64 command prompt - the prompt I am using is called "Developer Command Prompt" running on my x64 machine.  Not really sure how to check this.  I also tried the "x64 Native Tools Command Prompt", which gets the same error. Here is the banner from the command prompt
** Visual Studio 2022 Developer Command Prompt v17.9.5
** Copyright (c) 2022 Microsoft Corporation

2. I am making the makefile with cmake as described above.  I have not opened the project in Visual Studio (though I could try that if you think it is a good idea)
3. I just selected the express config, not sure if that installed the VS extension or not.  I just reinstalled using custom and made sure that the VS extension was turned on.  I still get the same error when running cmake on the command line with the VS extension installed (this is with both teh Developer Command Prompt and the x64 Native Tools Command Prompt

c:\Users\petew\source\repos\cuda_nbody\build>cmake .. -DCMAKE_BUILD_TYPE=Release
-- Running vcpkg install
...

**CMake Error at C:/Program Files/CMake/share/cmake-3.29/Modules/CMakeDetermineCompilerId.cmake:563 (message):
  No CUDA toolset found.**

-- Configuring incomplete, errors occurred!



**Below is the contents of CMakeConfigureLog.yaml:**

---
events:
  -
    kind: "message-v1"
    backtrace:
      - "C:/Program Files/CMake/share/cmake-3.29/Modules/CMakeDetermineSystem.cmake:205 (message)"
      - "CMakeLists.txt:15 (project)"
    message: |
      The system is: Windows - 10.0.22631 - AMD64
  -
    kind: "message-v1"
    backtrace:
      - "C:/Program Files/CMake/share/cmake-3.29/Modules/CMakeDetermineCompilerId.cmake:17 (message)"
      - "C:/Program Files/CMake/share/cmake-3.29/Modules/CMakeDetermineCompilerId.cmake:64 (__determine_compiler_id_test)"
      - "C:/Program Files/CMake/share/cmake-3.29/Modules/CMakeDetermineCCompiler.cmake:123 (CMAKE_DETERMINE_COMPILER_ID)"
      - "CMakeLists.txt:15 (project)"
    message: |
      Compiling the C compiler identification source file "CMakeCCompilerId.c" succeeded.
      Compiler:  
      Build flags: 
      Id flags:  
      
      The output was:
      0
      MSBuild version 17.9.8+b34f75857 for .NET Framework
      Build started 7/27/2024 3:11:34 PM.
      
      Project "C:\\Users\\petew\\source\\repos\\cuda_nbody\\build\\CMakeFiles\\3.29.0-rc3\\CompilerIdC\\CompilerIdC.vcxproj" on node 1 (default targets).
      PrepareForBuild:
        Creating directory "Debug\\".
        Structured output is enabled. The formatting of compiler diagnostics will reflect the error hierarchy. See https://aka.ms/cpp/structured-output for more details.
        Creating directory "Debug\\CompilerIdC.tlog\\".
      InitializeBuildStatus:
        Creating "Debug\\CompilerIdC.tlog\\unsuccessfulbuild" because "AlwaysCreate" was specified.
        Touching "Debug\\CompilerIdC.tlog\\unsuccessfulbuild".
      ClCompile:
        C:\\Program Files (x86)\\Microsoft Visual Studio\\2022\\BuildTools\\VC\\Tools\\MSVC\\14.39.33519\\bin\\HostX64\\x64\\CL.exe /c /nologo /W0 /WX- /diagnostics:column /Od /D _MBCS /Gm- /EHsc /RTC1 /MDd /GS /fp:precise /Zc:wchar_t /Zc:forScope /Zc:inline /Fo"Debug\\\\" /Fd"Debug\\vc143.pdb" /external:W0 /Gd /TC /FC /errorReport:queue CMakeCCompilerId.c
        CMakeCCompilerId.c
      Link:
        C:\\Program Files (x86)\\Microsoft Visual Studio\\2022\\BuildTools\\VC\\Tools\\MSVC\\14.39.33519\\bin\\HostX64\\x64\\link.exe /ERRORREPORT:QUEUE /OUT:".\\CompilerIdC.exe" /INCREMENTAL:NO /NOLOGO kernel32.lib user32.lib gdi32.lib winspool.lib comdlg32.lib advapi32.lib shell32.lib ole32.lib oleaut32.lib uuid.lib odbc32.lib odbccp32.lib /MANIFEST /MANIFESTUAC:"level='asInvoker' uiAccess='false'" /manifest:embed /PDB:".\\CompilerIdC.pdb" /SUBSYSTEM:CONSOLE /TLBID:1 /DYNAMICBASE /NXCOMPAT /IMPLIB:".\\CompilerIdC.lib" /MACHINE:X64 Debug\\CMakeCCompilerId.obj
        CompilerIdC.vcxproj -> C:\\Users\\petew\\source\\repos\\cuda_nbody\\build\\CMakeFiles\\3.29.0-rc3\\CompilerIdC\\CompilerIdC.exe
      PostBuildEvent:
        for %%i in (cl.exe) do @echo CMAKE_C_COMPILER=%%~$PATH:i
        :VCEnd
        CMAKE_C_COMPILER=C:\\Program Files (x86)\\Microsoft Visual Studio\\2022\\BuildTools\\VC\\Tools\\MSVC\\14.39.33519\\bin\\Hostx64\\x64\\cl.exe
      FinalizeBuildStatus:
        Deleting file "Debug\\CompilerIdC.tlog\\unsuccessfulbuild".
        Touching "Debug\\CompilerIdC.tlog\\CompilerIdC.lastbuildstate".
      Done Building Project "C:\\Users\\petew\\source\\repos\\cuda_nbody\\build\\CMakeFiles\\3.29.0-rc3\\CompilerIdC\\CompilerIdC.vcxproj" (default targets).
      
      Build succeeded.
          0 Warning(s)
          0 Error(s)
      
      Time Elapsed 00:00:00.92
      
      
      Compilation of the C compiler identification source "CMakeCCompilerId.c" produced "CompilerIdC.exe"
      
      Compilation of the C compiler identification source "CMakeCCompilerId.c" produced "CompilerIdC.vcxproj"
      
      The C compiler identification is MSVC, found in:
        C:/Users/petew/source/repos/cuda_nbody/build/CMakeFiles/3.29.0-rc3/CompilerIdC/CompilerIdC.exe
      
  -
    kind: "message-v1"
    backtrace:
      - "C:/Program Files/CMake/share/cmake-3.29/Modules/CMakeDetermineCompilerId.cmake:17 (message)"
      - "C:/Program Files/CMake/share/cmake-3.29/Modules/CMakeDetermineCompilerId.cmake:64 (__determine_compiler_id_test)"
      - "C:/Program Files/CMake/share/cmake-3.29/Modules/CMakeDetermineCXXCompiler.cmake:126 (CMAKE_DETERMINE_COMPILER_ID)"
      - "CMakeLists.txt:15 (project)"
    message: |
      Compiling the CXX compiler identification source file "CMakeCXXCompilerId.cpp" succeeded.
      Compiler:  
      Build flags: 
      Id flags:  
      
      The output was:
      0
      MSBuild version 17.9.8+b34f75857 for .NET Framework
      Build started 7/27/2024 3:11:35 PM.
      
      Project "C:\\Users\\petew\\source\\repos\\cuda_nbody\\build\\CMakeFiles\\3.29.0-rc3\\CompilerIdCXX\\CompilerIdCXX.vcxproj" on node 1 (default targets).
      PrepareForBuild:
        Creating directory "Debug\\".
        Structured output is enabled. The formatting of compiler diagnostics will reflect the error hierarchy. See https://aka.ms/cpp/structured-output for more details.
        Creating directory "Debug\\CompilerIdCXX.tlog\\".
      InitializeBuildStatus:
        Creating "Debug\\CompilerIdCXX.tlog\\unsuccessfulbuild" because "AlwaysCreate" was specified.
        Touching "Debug\\CompilerIdCXX.tlog\\unsuccessfulbuild".
      ClCompile:
        C:\\Program Files (x86)\\Microsoft Visual Studio\\2022\\BuildTools\\VC\\Tools\\MSVC\\14.39.33519\\bin\\HostX64\\x64\\CL.exe /c /nologo /W0 /WX- /diagnostics:column /Od /D _MBCS /Gm- /EHsc /RTC1 /MDd /GS /fp:precise /Zc:wchar_t /Zc:forScope /Zc:inline /Fo"Debug\\\\" /Fd"Debug\\vc143.pdb" /external:W0 /Gd /TP /FC /errorReport:queue CMakeCXXCompilerId.cpp
        CMakeCXXCompilerId.cpp
      Link:
        C:\\Program Files (x86)\\Microsoft Visual Studio\\2022\\BuildTools\\VC\\Tools\\MSVC\\14.39.33519\\bin\\HostX64\\x64\\link.exe /ERRORREPORT:QUEUE /OUT:".\\CompilerIdCXX.exe" /INCREMENTAL:NO /NOLOGO kernel32.lib user32.lib gdi32.lib winspool.lib comdlg32.lib advapi32.lib shell32.lib ole32.lib oleaut32.lib uuid.lib odbc32.lib odbccp32.lib /MANIFEST /MANIFESTUAC:"level='asInvoker' uiAccess='false'" /manifest:embed /PDB:".\\CompilerIdCXX.pdb" /SUBSYSTEM:CONSOLE /TLBID:1 /DYNAMICBASE /NXCOMPAT /IMPLIB:".\\CompilerIdCXX.lib" /MACHINE:X64 Debug\\CMakeCXXCompilerId.obj
        CompilerIdCXX.vcxproj -> C:\\Users\\petew\\source\\repos\\cuda_nbody\\build\\CMakeFiles\\3.29.0-rc3\\CompilerIdCXX\\CompilerIdCXX.exe
      PostBuildEvent:
        for %%i in (cl.exe) do @echo CMAKE_CXX_COMPILER=%%~$PATH:i
        :VCEnd
        CMAKE_CXX_COMPILER=C:\\Program Files (x86)\\Microsoft Visual Studio\\2022\\BuildTools\\VC\\Tools\\MSVC\\14.39.33519\\bin\\Hostx64\\x64\\cl.exe
      FinalizeBuildStatus:
        Deleting file "Debug\\CompilerIdCXX.tlog\\unsuccessfulbuild".
        Touching "Debug\\CompilerIdCXX.tlog\\CompilerIdCXX.lastbuildstate".
      Done Building Project "C:\\Users\\petew\\source\\repos\\cuda_nbody\\build\\CMakeFiles\\3.29.0-rc3\\CompilerIdCXX\\CompilerIdCXX.vcxproj" (default targets).
      
      Build succeeded.
          0 Warning(s)
          0 Error(s)
      
      Time Elapsed 00:00:00.81
      
      
      Compilation of the CXX compiler identification source "CMakeCXXCompilerId.cpp" produced "CompilerIdCXX.exe"
      
      Compilation of the CXX compiler identification source "CMakeCXXCompilerId.cpp" produced "CompilerIdCXX.vcxproj"
      
      The CXX compiler identification is MSVC, found in:
        C:/Users/petew/source/repos/cuda_nbody/build/CMakeFiles/3.29.0-rc3/CompilerIdCXX/CompilerIdCXX.exe
      
...

**Comment by tlemo** [comment]
Hi @petewil0 , the log looks incomplete. Can you please delete everything under the build directory, run the cmake again, then add the log file as an attachment to this thread?

**Comment by petewil0** [comment]
Had to rename the .yaml file to .txt to add it to github, so you may want to rename this back to .yaml
[CMakeConfigureLog.txt](https://github.com/user-attachments/files/16434341/CMakeConfigureLog.txt)

**Comment by petewil0** [comment]
(and yes, I deleted everything and did a clean build)

**Comment by tlemo** [comment]
Something doesn't make sense. The attached log has no mention of the CUDA configuration. At very least, it should contain something like:

```
...
    kind: "message-v1"
    backtrace:
      - "cmake/share/cmake-3.26/Modules/CMakeDetermineCompilerId.cmake:1102 (message)"
      - "cmake/share/cmake-3.26/Modules/CMakeDetermineCUDACompiler.cmake:75 (CMAKE_DETERMINE_COMPILER_ID_VENDOR)"
...
```

Do you use a modified/custom cmake? If so, please try to install the official cmake too.

You can also also try this minimal example of a CUDA + CMake setup: https://github.com/tlemo/hello_cuda

**Comment by petewil0** [comment]
I have two installations of cmake on my machine.  
C:\Program Files\CMake\bin\cmake.exe
C:\Program Files (x86)\Microsoft Visual Studio\2022\BuildTools\Common7\IDE\CommonExtensions\Microsoft\CMake\CMake\bin\cmake.exe

When I run cmake --version, the first one (retail cmake) gives me:
cmake version 3.29.0-rc3

When I run on the one from visual studio, I get the following version string:
cmake version 3.28.0-msvc1


If I just run cmake --version from the Developer Command Prompt, I get version 3.29.0-rc3.
So, I think I am running the correct cmake.

hello_cuda is also failing to configure, CMakeConfigureLog.txt added below
[CMakeConfigureLog.txt](https://github.com/user-attachments/files/16475619/CMakeConfigureLog.txt)



I did try to compile cmake from source at one point, so I tried uninstalling and reinstalling cmake.
The new version of cmake is: cmake version 3.30.1

After reinstalling, I cleaned the build directory for both hello_cuda and cuda_nbody, and I get what looks like the same error with hello_cuda (nvcc error : 'cudafe++' died with status 0xC0000005 (ACCESS_VIOLATION))

It also looks like I get the same error with cuda_nbody:
CMake Error at C:/Program Files/CMake/share/cmake-3.30/Modules/CMakeDetermineCompilerId.cmake:587 (message):
  No CUDA toolset found.

**Comment by tlemo** [comment]
> nvcc error : 'cudafe++' died with status 0xC0000005 (ACCESS_VIOLATION)

Ah, there was an incompatibility between CUDA 12.5 and some of the recent VS 2022 updates. If that is the root problem you should be able to resolve this by using CUDA 12.5.1 (or 12.6 which was just released)

**Comment by petewil0** [comment]
Thanks!
Updated Cuda to 12.6 - same failure
Updated VS22 to the very latest - same failure.
Deleted everything in "build" and retried, same failure

Any other ideas?

[CMakeConfigureLog.txt](https://github.com/user-attachments/files/16481878/CMakeConfigureLog.txt)

**Comment by petewil0** [comment]
(that was in hello_cuda, by the way, not cuda_nbody)

**Comment by tlemo** [comment]
The latest log attached end with:

```
    kind: "message-v1"
    backtrace:
      - "C:/Program Files/CMake/share/cmake-3.30/Modules/CMakeDetermineCompilerId.cmake:1192 (message)"
      - "C:/Program Files/CMake/share/cmake-3.30/Modules/CMakeDetermineCUDACompiler.cmake:75 (CMAKE_DETERMINE_COMPILER_ID_VENDOR)"
      - "CMakeLists.txt:4 (project)"
    message: |
      Checking whether the CUDA compiler is NVIDIA using "" matched "nvcc: NVIDIA \\(R\\) Cuda compiler driver":
      nvcc: NVIDIA (R) Cuda compiler driver
      Copyright (c) 2005-2024 NVIDIA Corporation
      Built on Fri_Jun_14_16:44:19_Pacific_Daylight_Time_2024
      Cuda compilation tools, release 12.6, V12.6.20
      Build cuda_12.6.r12.6/compiler.34431801_0
```

It appears it located the right nvcc, which errors did you see? Do you still get the `nvcc error : 'cudafe++' died with status 0xC0000005 (ACCESS_VIOLATION)`?

Can you please make sure you're using a amd64/x64 MSVC command prompt? Based on the latest log it seems you're using an 32bit (x86) command prompt. Also, make sure you're using a ninja or nmake generator - I believe the `No CUDA toolset found` error is only reported when you're trying to generate a VS solution.

If you're willing to do some cmake debugging, you could try adding messages in `CMakeDetermineCompilerId.cmake`  and figure out where/why exactly it fails.

Also, see this older tread on stackoverflow: https://stackoverflow.com/questions/56636714/cuda-compile-problems-on-windows-cmake-error-no-cuda-toolset-found

**Comment by petewil0** [comment]
Thanks again for getting back to me!

(this is all for hello_cuda)

1. yes, I was seeing "nvcc error : 'cudafe++' died with status 0xC0000005 (ACCESS_VIOLATION)
2. I'm not really sure which command prompt to use.  I'm using "Developer Command prompt", and not really sure whether it is set up for 32 or 64 (and I'm not sure how to tell, in fact).  **Which command prompt should I be using? (or what arguments to vcvarsall?**)  I'm on a x64 machine.
3. I tried all of the following, not sure which I'm supposed to use
    a. Developer Command Prompt - it says [vcvarsall.bat] Environment initialized for: 'x64', but you were saying it was 32 bit...
  It gives an error:   nvcc error : 'cudafe++' died with status 0xC0000005 (ACCESS_VIOLATION)
    b. x64 native tools command prompt - it says [vcvarsall.bat] Environment initialized for: 'x64' but gives an error:
    CMakeFiles\cmTC_04091.dir\main.cu.obj : fatal error LNK1112: module machine type 'x64' conflicts with target machine type 'x86'
    c. x86 native tools command prompt - It says "[vcvarsall.bat] Environment initialized for: 'x86', but give an error
  nvcc error : 'cudafe++' died with status 0xC0000005 (ACCESS_VIOLATION)
**I wonder why the x86 tools command prompt is getting initialized for x64, that seems odd.**
    d. x64_x86 Cross Tools Command Prompt: It says   [vcvarsall.bat] Environment initialized for: 'x64' and gets error: 
  nvcc error : 'cudafe++' died with status 0xC0000005 (ACCESS_VIOLATION)
    e. x86_x64 Cross Tools Command Prompt:  It says [vcvarsall.bat] Environment initialized for: 'x86_x64' and gets error: 
    CMakeFiles\cmTC_8953e.dir\main.cu.obj : fatal error LNK1112: module machine type 'x64' conflicts with target machine type 'x86'


Looking at the command line output, it seems that there is a recursive call to vcvarsall - in the x86 case, for instance, on the first call it says [vcvarsall.bat] Environment initialized for: 'x86', but later on in the output it says [vcvarsall.bat] Environment initialized for: 'x64'

4. I don't quite understand your suggestion to "make sure you're using a ninja or nmake generator ".  I did check and both ninja and nmake are installed, is that what you meant, or did you mean something else, like checking to make sure that the cmakefile.txt is properly calling ninja?

5. Sure, happy to do some cmake debugging, but new to it, so it may take me awhile to get results.
6. The old stack overflow thread seem to deal with the error in cuda_nbody, but not the error in hello_cuda.  Still, I'll give it a try and see if it fixes cuda_nbody

**Comment by tlemo** [comment]
1. The NVCC crash is unexpected. Please set the following env variable `NVCC_REPORT_ALLERROR=1` and do a clean reconfigure, noting any additional error messages

2. Regarding the generators, you can specify one of the target build system "generators" from the command line, ex `cmake -G Ninja ..`

**Comment by petewil0** [comment]
1. OK, here is a new error after adding NVCC_REPORT_ALLERROR=1 - it looks like there are about 71 errors of type invalid redeclaration of <foo>
  #$ cudafe++ --ms_c++14 --microsoft_version=1940 --msvc_target_version=1940
  --compiler_bindir "C:/Program Files (x86)/Microsoft Visual
  Studio/2022/BuildTools/VC/Tools/MSVC/14.40.33807/bin/HostX86/x86/../../../../../../.."
  --sdk_dir "C:/Program Files (x86)/Windows Kits/10/" --display_error_number
  --orig_src_file_name "CMakeCUDACompilerId.cu" --orig_src_path_name
  "C:\Users\petew\source\repos\hello_cuda\build\CMakeFiles\3.30.1\CompilerIdCUDA\CMakeCUDACompilerId.cu"
  --allow_managed --m64 --parse_templates --gen_c_file_name
  "tmp/CMakeCUDACompilerId.cudafe1.cpp" --stub_file_name
  "CMakeCUDACompilerId.cudafe1.stub.c" --gen_module_id_file
  --module_id_file_name "tmp/CMakeCUDACompilerId.module_id"
  "tmp/CMakeCUDACompilerId.cpp4.ii"

  C:\Program Files (x86)\Microsoft Visual
  Studio\2022\BuildTools\VC\Tools\MSVC\14.40.33807\include\vcruntime.h(192):
  error: invalid redeclaration of type name "size_t"

        typedef unsigned int     size_t;

I uninstalled and reinstalled cuda 12.5, 12.6, and VS 2022, reran, and got the same problem of error compiling while seeing previous definitions of things

2. Ah, yes, I was already using "-G Ninja" when executing the command for hello_cuda: I use this command to build/configure it
"cmake .. -G Ninja -DCMAKE_BUILD_TYPE=Release"

I think the next step is on my side to try debugging the cmakefiles as you suggested, it will take me a bit of time to learn how to debug cmake makefiles.

**Comment by tlemo** [comment]
> C:\Program Files (x86)\Microsoft Visual
Studio\2022\BuildTools\VC\Tools\MSVC\14.40.33807\include\vcruntime.h(192):
error: invalid redeclaration of type name "size_t"

Is this the only error? Same command that is crashing nvcc? (this error should not lead to a crash, unless there's a new nvcc bug surfacing here)

> Ah, yes, I was already using "-G Ninja" when executing the command for hello_cuda: I use this command to build/configure it
"cmake .. -G Ninja -DCMAKE_BUILD_TYPE=Release"

You still get the `No CUDA toolset found` error? 

>The old stack overflow thread seem to deal with the error in cuda_nbody, but not the error in hello_cuda.

I just noticed this comment: are you getting different errors from cuda_nbody vs the ones from hello_cuda? 

My approach for troubleshooting cmake scritps was simply injecting [message()](https://cmake.org/cmake/help/latest/command/message.html) tracing (aka printf debugging). But it seems that more civilized cmake-debugging solutions are starting to come up:

- https://devblogs.microsoft.com/cppblog/introducing-cmake-debugger-in-vs-code-debug-your-cmake-scripts-using-open-source-cmake-debugger
- https://devblogs.microsoft.com/cppblog/cmake-debugger-allows-you-to-debug-your-cmake-scripts-and-more/
- https://stackoverflow.com/questions/22803607/how-can-i-debug-cmakelists-txt-files

**Comment by petewil0** [comment]
**Hooray, I have some progress!  (however, things are now failing at the "make" step, more details below)  Is there supposed to be a "makefile" in cuda_nbody\build after I run the configure (cmake) step?**


1. Yes, the only error that I am seeing when I build hello_cuda  is the type name redeclaration error, and I am seeing 71 of them before I finally get the crash.
2. After I uninstalled Visual Studio, VS Command line tools, and CUDA 12.5 and 12.6 and reinstalled, I no longer get the CUDA toolset not found error when compiling cuda_nbody.  The cmake configure step succeeded!  When I tried briefly before cleaning the build directory, it looked like the previously configured cmakefiles were trying to use VS build tools (command line) instead of VS itself, so that may be related to where the problem came from.
3. Yes, I do get different errors from hello_cuda and cuda_nbody.  hello_cuda is getting the nvcc crash, and cuda_nbody was getting the "no cuda toolset found" (but now it is getting an error trying to find the makefile)
4. Since I have gotten further than I did before, I can now try the next step for cuda_nbody, which is "make -j24".  However this fails.  I installed make from gnu, and I get an error that "makefile" is not found.  Your instructions don't say that you need to install "make" on windows platforms, it might be good to add that for others who build in the future.
c:\Users\petew\source\repos\cuda_nbody\build>make -j24
make: *** No targets specified and no makefile found.  Stop.

looking in cuda_nbody\build, I don't see a "makefile" - **should there be one?**

 Directory of c:\Users\petew\source\repos\cuda_nbody\build
08/05/2024  01:14 PM    <DIR>          .
07/27/2024  03:10 PM    <DIR>          ..
08/05/2024  01:14 PM            60,681 ALL_BUILD.vcxproj
08/05/2024  01:14 PM               293 ALL_BUILD.vcxproj.filters
08/05/2024  01:14 PM            21,305 CMakeCache.txt
08/05/2024  01:14 PM    <DIR>          CMakeFiles
08/05/2024  01:14 PM             1,768 cmake_install.cmake
08/05/2024  01:14 PM             3,110 nbody.sln
08/05/2024  01:14 PM            82,954 nbody.vcxproj
08/05/2024  01:14 PM             2,006 nbody.vcxproj.filters
08/05/2024  01:14 PM             3,971 vcpkg-manifest-install.log
08/05/2024  01:14 PM    <DIR>          vcpkg_installed
08/05/2024  01:14 PM            60,753 ZERO_CHECK.vcxproj
08/05/2024  01:14 PM               536 ZERO_CHECK.vcxproj.filters
              10 File(s)        237,377 bytes
               4 Dir(s)  334,784,516,096 bytes free

**Comment by petewil0** [comment]
Quick correction - in hello_cuda I am *not* seeing the nvcc crash anymore - just the 71 compilation errors, and then configure incomplete message:


  71 errors detected in the compilation of "CMakeCUDACompilerId.cu".

    # --error 0x2 --





Call Stack (most recent call first):
  C:/Program Files/CMake/share/cmake-3.30/Modules/CMakeDetermineCompilerId.cmake:8 (CMAKE_DETERMINE_COMPILER_ID_BUILD)
  C:/Program Files/CMake/share/cmake-3.30/Modules/CMakeDetermineCompilerId.cmake:53 (__determine_compiler_id_test)
  C:/Program Files/CMake/share/cmake-3.30/Modules/CMakeDetermineCUDACompiler.cmake:131 (CMAKE_DETERMINE_COMPILER_ID)
  CMakeLists.txt:4 (project)


-- Configuring incomplete, errors occurred!

**Comment by tlemo** [comment]
Can you please share the full output (with the 71 errors) & the CMakeConfigureLog.yaml file?

**Comment by petewil0** [comment]
[hello_cuda_errors.txt](https://github.com/user-attachments/files/16515695/hello_cuda_errors.txt)
[CMakeConfigureLog.txt](https://github.com/user-attachments/files/16515700/CMakeConfigureLog.txt)

Here you go!

**Comment by petewil0** [comment]
OK, I finally got back to this and got to the next step.  

I think the problem was that the CUDA compiler was trying to use some 32 bit tools for a 64 bit build because I had a bad path set somewhere.

It looks like I had a combination of problems - At one point I used a developer command prompt for cross compiling (maybe the one called "Developer Command Prompt for VS 2022"), and that left the build directory in an inconsistent state.

I solved the problems by first deleting the build directory, then opening a different command prompt shell - "x64 Native Tools Command Prompt for VS 2022", and rebuilding. (this should be equivalent to running vsvarsall.bat x64)

With that specific command shell, both hello_cuda and cuda_nbody successfully get past the cmake step.
hello_cuda now builds and runs for me.  (I'll need to install "make" on my windows box to try cuda_nbody).

**Comment by petewil0** [comment]
And I got it to run - one quick update you might want to make to the instructions.  On your github page, you say that to build it we should do:
mkdir build && cd build
cmake .. -DCMAKE_BUILD_TYPE=Release
make -j8 # replace 8 with the number of cores

However, there is no makefile to run the "make -j8" on, at least not on windows.  Instead, I opened the project in visual studio, and I was able to build it from there.  While it didn't run for me, I was able to run by double clicking on the built executable.

So, I think I have everything figured out and working now, closing this issue.
